src/utils/hashlinkEncoder.ts

```diff
@@ -0,0 +1,124 @@
+import { MultihashEncoder } from './MultihashEncoder'
+
+type Metadata = {
+  urls?: string[]
+  contentType?: string
+}
+
+export type HashlinkData = {
+  checksum: string
+  metadata?: Metadata
+}
+
+const URLS = 0x0f
+const CONTENTTYPE = 0x0e
+
+export class HashlinkEncoder {
+  /**
+   *
+   * Encodes a buffer, with optional metadata, into a hashlink
+   *
+   * @param {Buffer} buffer the buffer to encode into a hashlink
+   * @param {string} hashName the name of the hashing algorithm 'sha2-256'
+   * @param {BaseName} baseName the name of the base encoding algorithm 'base58btc'
+   * @param {Metadata} metadata the optional metadata in the hashlink
+   *
+   * @returns {string} hashlink hashlink with optional metadata
+   */
+  public static encode(buffer: Buffer, hashName: 'sha2-256', baseName: BaseName = 'base58btc', metadata?: Metadata) {
+    const checksum = this.encodeMultihashEncoder(buffer, hashName, baseName)
+    const mbMetadata = metadata ? this.encodeMetadata(metadata, baseName) : null
+    return mbMetadata ? `hl:${checksum}:${mbMetadata}` : `hl:${checksum}`
+  }
+
+  /**
+   *
+   * Decodes a hashlink into HashlinkData object
+   *
+   * @param {string} hashlink the hashlink that needs decoding
+   *
+   * @returns {HashlinkData} object the decoded hashlink
+   */
+  public static decode(hashlink: string): HashlinkData {
+    const hashlinkList = hashlink.split(':')
+    if (this.validate(hashlink)) {
+      const checksum = hashlinkList[1]
+      const metadata = hashlinkList[2] ? this.decodeMetadata(hashlinkList[2]) : null
+
+      return metadata ? { checksum, metadata } : { checksum }
+    } else {
+      throw new Error(`Invalid hashlink: ${hashlink}`)
+    }
+  }
+
+  public static validate(hashlink: string): boolean {
+    const hashlinkList = hashlink.split(':')
+    const validMultibase = MultibaseEncoder.validate(hashlinkList[1])
+    if (!validMultibase) {
+      return false
+    }
+    const { data } = MultibaseEncoder.decode(hashlinkList[1])
+    const validMultihash = MultihashEncoder.validate(data)
+    return validMultibase && validMultihash ? true : false
+  }
+
+  private static encodeMultihashEncoder(buffer: Buffer, hashName: 'sha2-256', baseName: BaseName): string {
+    // TODO: Support more hashing algorithms
+    const hash = new Uint8Array(sha256.array(buffer))
+    const mh = MultihashEncoder.encode(hash, hashName)
+    const mb = MultibaseEncoder.encode(mh, baseName)
+    return BufferEncoder.toUtf8String(mb)
+  }
+
+  private static encodeMetadata(metadata: Metadata, baseName: BaseName): string {
+    const metadataMap = new Map()
+
+    for (const key of Object.keys(metadata)) {
+      switch (key) {
+        case 'urls':
+          metadataMap.set(URLS, metadata.urls)
+          break
+        case 'contentType':
+          metadataMap.set(CONTENTTYPE, metadata.contentType)
+          break
+        default:
+          throw new Error(`Invalid metadata: ${metadata}`)
+      }
+    }
```

Comment:
I'm thinking about how to get rid of the switch statement and something like this could work:
```js

hexTable = {
  urls: 0x0f, 
  contentType: 0x0e
}

for (const key of Object.keys(metadata)) {
  if (hexTable[key]) { 
    metadataMap.set(hexTable[key], metadata[key])
  } else {
    throw new Error(...)
  }
}
```

Reply:
Similarily for the `decodeMetadata`. But maybe I'm missing something which would complicate the code anyway 🤷‍♂️